devito/function.py

```diff
@@ -335,16 +331,6 @@ def _extent_padding_left(self):
         """
         return self._offset_halo
 
-    def initialize(self):
```

Comment:
I don't think this should be removed

Reply:
why? 
unused and basically useless => drop

Reply:
It was being used in tests that you modify in this same PR. As I mention, they are quite useful when you're using an external file reader to read in the data. They are also useful in ensuring that the memory isn't allocated too early in the workflow only to specify initialisation. With initializers we can delay memory allocation as much as we want. IIRC Yask has a similar concept.

Reply:
I didn't get the comment about delayed memory allocation (memory is lazily allocated already), but fine, I can revert it

Reply:
Lazy allocation is pointless if initialisation triggers memory allocation. With the initializer in place, memory allocation is deferred further until just before the operator is ready to execute.

Reply:
> Lazy allocation is pointless if initialisation triggers memory allocation.

Assuming that by "initialisation" you mean instantiation of the object, then memory is not allocated straight away; it is allocated the first time one tries to access ``.data``. 

That's why I was droppining the initialiser:
```
f = Function(... initialier=...)
```
is identical to
```
f = Function(...)
f.data[...] = ...
```

Reply:
> memory allocation is deferred further until just before the operator is ready to execute.

Redundant comment: memory allocation is not strictly related to an Operator per se, but rather to the first access to ``.data``. The fact that you mention an Operator is because clearly an Operator has to access ``.data``

Reply:
No. By initialisation, I mean initialisation, i.e. I instantiate by `u = TimeData()`, and I initialise with `u.data[:] = ...`. With the initializer, I can provide the initialisation _routine_ without calling it, so it can be called later in the workflow, hence making memory allocation lazier - this way we hold memory for less time.

Reply:
oh yes, in that sense I agree. As I said somewhere else, I'll put ``initialiser`` back to its place